Applications/Spire/Source/Ui/KeyFilterPanel.cpp

```diff
@@ -0,0 +1,103 @@
+#include "Spire/Ui/KeyFilterPanel.hpp"
+#include "Spire/Ui/AnyInputBox.hpp"
+#include "Spire/Ui/CustomQtVariants.hpp"
+#include "Spire/Ui/KeyInputBox.hpp"
+
+using namespace boost::signals2;
+using namespace Spire;
+
+struct KeyInputBoxValueModel : ValueModel<QKeySequence> {
+  std::shared_ptr<KeySequenceValueModel> m_source;
+  std::shared_ptr<AnyListModel> m_matches;
+  std::unordered_set<QString> m_matches_set;
+  std::vector<QString> m_matches_list;
+  scoped_connection m_matches_connection;
+
+  KeyInputBoxValueModel(std::shared_ptr<KeySequenceValueModel> source,
+      std::shared_ptr<AnyListModel> matches)
+      : m_source(std::move(source)),
+        m_matches(std::move(matches)),
+        m_matches_connection(m_matches->connect_operation_signal(
+          std::bind_front(&KeyInputBoxValueModel::on_operation, this))) {
+    for(auto i = 0; i < m_matches->get_size(); ++i) {
+      add_match(i);
+    }
+  }
+
+  QValidator::State get_state() const override {
+    return m_source->get_state();
+  }
+
+  const Type& get() const override {
+    return m_source->get();
+  }
+
+  QValidator::State test(const Type& value) const {
+    return m_source->test(value);
+  }
+
+  QValidator::State set(const Type& value) {
+    if(m_matches_set.contains(value.toString())) {
+      return QValidator::State::Invalid;;
+    }
+    return m_source->set(value);;
+  }
+
+  connection connect_update_signal(
+      const UpdateSignal::slot_type& slot) const override {
+    return m_source->connect_update_signal(slot);
+  }
+
+  void on_operation(const AnyListModel::Operation& operation) {
+    visit(operation,
+      [&] (const AnyListModel::AddOperation& operation) {
+        add_match(operation.m_index);
+      },
+      [&] (const AnyListModel::RemoveOperation& operation) {
+        m_matches_set.erase(m_matches_list[operation.m_index]);
+        m_matches_list.erase(m_matches_list.begin() + operation.m_index);
+      });
+  }
+
+  void add_match(int index) {
+    auto value = displayTextAny(m_matches->get(index));
+    m_matches_set.insert(value);
+    m_matches_list.insert(m_matches_list.begin() + index, value);
```

Comment:
What does ```m_matches_list``` provide that ```m_matches``` does not?

Reply:
When `KeyInputBoxValueModel` gets the `RemoveOperation` signal, the item has been removed from `m_matches`, and `m_matches_set` doesn't know which item should be removed. `m_matches_list` helps `m_matches_set` to quickly remove items.